src/util/reference_processor.rs

```diff
@@ -129,207 +162,392 @@ pub enum Semantics {
-
-        // FIXME: We are calling mutator() for a worker thread
-        panic!("We are calling mutator() for a worker tls. We need to fix this.");
-        // unsafe { VM::VMActivePlan::mutator(tls)) }.flush_remembered_sets();
-        // if TRACE {
-        //     trace!("Ending ReferenceProcessor.scan({:?})", self.semantics);
-        // }
+        debug!("Ending ReferenceProcessor.scan({:?})", self.semantics);
     }
 
-    /**
-     * This method deals only with soft references. It retains the referent
-     * if the reference is definitely reachable.
-     * @param reference the address of the reference. This may or may not
-     * be the address of a heap object, depending on the VM.
-     * @param trace the thread local trace element.
-     */
-    fn retain_referent<VM: VMBinding, T: TraceLocal>(
-        &self,
-        trace: &mut T,
-        reference: ObjectReference,
-    ) {
+    /// This method deals only with soft references. It retains the referent
+    /// if the reference is definitely reachable.
+    fn retain_referent<E: ProcessEdgesWork>(&self, trace: &mut E, reference: ObjectReference) {
         debug_assert!(!reference.is_null());
         debug_assert!(self.semantics == Semantics::SOFT);
 
-        if TRACE_DETAIL {
-            trace!("Processing reference: {:?}", reference);
-        }
+        trace!("Processing reference: {:?}", reference);
 
         if !reference.is_live() {
-            /*
-             * Reference is currently unreachable but may get reachable by the
-             * following trace. We postpone the decision.
-             */
+            // Reference is currently unreachable but may get reachable by the
+            // following trace. We postpone the decision.
             return;
         }
 
-        /*
-         * Reference is definitely reachable.  Retain the referent.
-         */
-        let referent = VM::VMReferenceGlue::get_referent(reference);
+        // Reference is definitely reachable.  Retain the referent.
+        let referent = <E::VM as VMBinding>::VMReferenceGlue::get_referent(reference);
         if !referent.is_null() {
-            trace.retain_referent(referent);
+            Self::keep_referent_alive(trace, referent);
         }
-        if TRACE_DETAIL {
-            trace!(" ~> {:?} (retained)", referent.to_address());
+        trace!(" ~> {:?} (retained)", referent.to_address());
+    }
+
+    /// Process a reference.
+    /// * If both the reference and the referent is alive, return the updated reference and update its referent properly.
+    /// * If the reference is alive, and the referent is not null but not alive, return a null pointer and the reference (with cleared referent) is enqueued.
+    /// * For other cases, return a null pointer.
+    ///
+    /// If a null pointer is returned, the reference can be removed from the reference table. Otherwise, the updated reference should be kept
+    /// in the reference table.
```

Comment:
It is more idiomatic in Rust to return `Option<ObjectReference>` instead.  We never store the returned `NULL` objref in any ref fields.  We only use the return value to decide if the reference should be discarded or retained.

Reply:
Sounds good.

Reply:
With this change, we are able to use a more concise way to iterate on the reference table `references.iter().filter_map(|reff| self.process_reference(...))`.